client/internal/statemanager/manager.go

```diff
@@ -0,0 +1,298 @@
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/exp/maps"
+
+	nberrors "github.com/netbirdio/netbird/client/errors"
+)
+
+// State interface defines the methods that all state types must implement
+type State interface {
+	Name() string
+	Cleanup() error
+}
+
+// Manager handles the persistence and management of various states
+type Manager struct {
+	mu     sync.Mutex
+	cancel context.CancelFunc
+	done   chan struct{}
+
+	filePath string
+	// holds the states that are registered with the manager and that are to be persisted
+	states map[string]State
+	// holds the state names that have been updated and need to be persisted with the next save
+	dirty map[string]struct{}
+	// holds the type information for each registered state
+	stateTypes map[string]reflect.Type
+}
+
+// New creates a new Manager instance
+func New(filePath string) *Manager {
+	return &Manager{
+		filePath:   filePath,
+		states:     make(map[string]State),
+		dirty:      make(map[string]struct{}),
+		stateTypes: make(map[string]reflect.Type),
+	}
+}
+
+// Start starts the state manager periodic save routine
+func (m *Manager) Start() {
+	if m == nil {
+		return
+	}
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	var ctx context.Context
+	ctx, m.cancel = context.WithCancel(context.Background())
+	m.done = make(chan struct{})
+
+	go m.periodicStateSave(ctx)
+}
+
+func (m *Manager) Stop(ctx context.Context) error {
+	if m == nil {
+		return nil
+	}
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if m.cancel != nil {
+		m.cancel()
+
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-m.done:
+			return nil
+		}
+	}
+
+	return nil
+}
+
+// RegisterState registers a state with the manager but doesn't attempt to persist it.
+// Pass an uninitialized state to register it.
+func (m *Manager) RegisterState(state State) {
+	if m == nil {
+		return
+	}
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	name := state.Name()
+	m.states[name] = nil
+	m.stateTypes[name] = reflect.TypeOf(state).Elem()
+}
+
+// GetState returns the state for the given type
+func (m *Manager) GetState(state State) State {
+	if m == nil {
+		return nil
+	}
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	return m.states[state.Name()]
+}
+
+// UpdateState updates the state in the manager and marks it as dirty for the next save.
+// The state will be replaced with the new one.
+func (m *Manager) UpdateState(state State) error {
+	if m == nil {
+		return nil
+	}
+
+	return m.setState(state.Name(), state)
+}
+
+// DeleteState removes the state from the manager and marks it as dirty for the next save.
+// Pass an uninitialized state to delete it.
+func (m *Manager) DeleteState(state State) error {
+	if m == nil {
+		return nil
+	}
+
+	return m.setState(state.Name(), nil)
+}
+
+func (m *Manager) setState(name string, state State) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if _, exists := m.states[name]; !exists {
+		return fmt.Errorf("state %s not registered", name)
+	}
+
+	m.states[name] = state
```

Comment:
if state is nil then we should not delete it from m.state[name] instead of set to nil?

Reply:
No, we still need it for the registration check (minor thing) and to (over)write  an empty state for `name` in the json with